core/src/main/scala/filodb.core/memstore/IndexMetadataStore.scala

```diff
@@ -86,6 +91,249 @@ trait IndexMetadataStore {
+   *  and synced, it is important to ensure the gen file is updated with the current generation. This guarantees that
+   *  next time the on bootstrap, the initState should return Synced and not TriggerRebuild
+   *
+   * @param datasetRef The dataset ref
+   * @param shard the shard id of index
+   * @param state One of the IndexState.Values for the index
+   * @param time a time in millis since epoch for when the state was updated
+   */
+  // scalastyle:off method.length
+  override def updateState(datasetRef: DatasetRef, shard: Int, state: IndexState.Value, time: Long): Unit = {
+    state match {
+      case IndexState.Synced           =>  // Only handle cases where state is Synced
+                                    val snapFile = new File(snapFilePathFormat.format(datasetRef.dataset, shard))
+                                    val genFile = new File(genFilePathFormat.format(datasetRef.dataset, shard))
+                                    Using(new FileOutputStream(snapFile)){
+                                      // First write the last synced time to the snap file
+                                      snapFos =>
+                                        logger.info("Updating state of dataset={}, shard={} with syncTime={}",
+                                          datasetRef.toString, shard, time)
+                                        snapFos.write(snapFileV1Magic)
+                                        val buff = ByteBuffer.allocate(8)
+                                        buff.putLong(time)
+                                        snapFos.write(buff.array())
+                                    }.flatMap { _ =>
+                                      // If writing the snap file is successful, update the generation file so
+                                      // next call to initState does not trigger the rebuild. Note that these
+                                      // two operations are not atomic, we may still have the snap file updated but
+                                      // generation file not, in which case next call to initState will trigger
+                                      // the rebuild.
+                                      expectedGeneration.map{
+                                        generation =>
+                                          Using(new FileOutputStream(genFile)) { genFos =>
+                                            logger.info("Updating genFile " +
+                                              "of dataset={}, shard={} with generation={}",
+                                              datasetRef.toString, shard, generation)
+                                            genFos.write(genFileV1Magic)
+                                            val buff = ByteBuffer.allocate(4)
+                                            buff.putInt(generation)
+                                            genFos.write(buff.array())
```

Comment:
separate function and UT for genFile write  ? ( UT would call this block and checks if it is written to file correctly )

Reply:
Same as above